src/msg/simple/Pipe.cc

```diff
@@ -2452,25 +2468,32 @@ int Pipe::tcp_read(char *buf, unsigned len)
       }
     }
 
-    if (tcp_read_wait() < 0)
-      return -1;
+    if (tcp_read_wait() < 0) {
+      rc = -errno;
```

Comment:
reading errno should be moved to right after the syscall, not in any of the callers of these wrappers. that'll avoid as much as possible getting invalid errno values, and avoiding accidental resetting of it if the code changes